### ISSUE TITLE
Tray is black after hide/show

### ISSUE THREAD

**Issue by magnetophon** [polybar/polybar#1995]
## Describe the issue
When I do ``polybar-msg cmd hide`` followed by ``polybar-msg cmd show``, the tray turns black and doesn't respond to clicks or hoovers anymore.

When I switch workspaces, it turns back to normal. 

It only happens with ``bottom = true``.

**Was it working before?**
I'm new to polybar, so it's always been like this for me.

## To Reproduce
```dosini
[bar/bar1]
enable-ipc = true
bottom = true
modules-center = my-text-label
tray-position = right

[module/my-text-label]
type = custom/text
content = Some random label
```
1) make sure you have icons in your tray
2) ``polybar-msg cmd hide``
3) ``polybar-msg cmd show`` 

## Polybar Log
```
polybar bar1
warn: No fonts specified, using fallback font "fixed"
```

## Environment:
* WM: i3
* Distro: NixOS
* Output of `polybar -vvv`:
```
polybar 3.4.2

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard

X extensions: +randr (+monitors) +composite +xkb +xrm +xcursor

Build type: Release
Compiler: gcc-wrapper-9.2.0/bin/g++
Compiler flags:  -Wall -Wextra -Wpedantic -O3 -DNDEBUG
Linker flags:
```

Thanks for this awesome bar!

**Comment by patrick96** [comment]
This sounds similar to #789 but there it seemed to happen randomly.

Does this happen consistently?

Can you make a screenshot when this happens?

**Comment by magnetophon** [comment]
Yes, this is 100% consistent.
Looks like this: https://imgur.com/a/4Fqf39M

**Comment by patrick96** [comment]
I was also able to reproduce this now on `Xephyr`. This only seems to happen on i3 because it doesn't happen on bspwm or herbstluftwm. Also, after the icon turns black, when you switch to another workspace and the icon appears again, polybar doesn't print any log messages, so it didn't even redraw the icon, it just magically appeared again.

**Comment by magnetophon** [comment]
Great, confirmed.
Thanks for looking in to this!

**Comment by patrick96** [comment]
So after quite a bit of digging I figured out what is happening.

For one the problem is that we don't restack the tray over the bar window after the bar is shown again, so the tray icons suddenly end up behind the polybar window, changing the workspace fixes that.

But even if I add code to restack the tray we have another issue:

If you have `override-redirect = false` (default), the tray restacking just fails because of the issues described in #1355 and #425. 

With `override-redirect = true` the tray for some reason doesn't get notified that is has been hidden by i3 and so doesn't know it's hidden and doesn't show itself again when the bar is shown.

The case without override-redirect should be easily fixable once I get around to finishing #1435 

The case with override-redirect is a little stranger. There is a dirty fix to just tell the tray when it has been unmapped instead of waiting for the window manager to tell us. I don't know what kind of side-effects this would have though. Also, I see no reason why the behavior should be different with or without override-redirect. Because the tray **always** has override-redirect enabled, even if the bar doesn't, so I don't understand why it depends on the bar's override-redirect setting whether `UnmapNotify` events are delivered to the tray window since the bar and tray window are completely independent of each other.

If there is someone who is more familiar with the X protocol, I'd appreciate your input, especially on the second point.

**Comment by somospocos** [comment]
@patrick96 I am a total ignoramus when it comes to X protocol and C/C++ programming, but I have witnessed this exact "ghost systray" syndrome in another software.

There are systray patches for different versions of dwm here https://dwm.suckless.org/patches/systray/ . While the https://dwm.suckless.org/patches/systray/dwm-6.1-systray.diff patch has worked flawlessly for years, the https://dwm.suckless.org/patches/systray/dwm-systray-20190208-cb3f58a.diff patch, which is supposed to be applied to dwm-6.2 , randomly causes the exact behavior reported by OP of this issue: "the tray turns black and doesn't respond to clicks or hoovers anymore". Perhaps you could have a look at those two patches and their diff, maybe it could provide any insights to you.

Unfortunately I have nothing more to add to this, because I never tried to get in touch with dwm developers and find any details, because I obviously don't meet their criteria of an advanced user that is described here https://suckless.org/philosophy/ , so I wouldn't have what to tell them in addition to "this doesn't work for me".

**Comment by magnetophon** [comment]
Update: it's not consistent here either.

**Comment by patrick96** [comment]
@somospocos Thanks. I will see what they do differently.

@magnetophon Really? I can reproduce it everytime.

**Comment by magnetophon** [comment]
Yeah, sometimes it stays visible for a while. I can toggle the bar as many times as I want and it never misbehaves.

Most of the time I get black after toggle. (actually, I have a semi-transparent bar now, so I don't get a black tray but a transparent one)

**Comment by patrick96** [comment]
Hmm weird. Maybe I am doing something differently. Because I basically only started the bar, ran `hide` and `show` and then killed the bar again. Maybe if it's running longer the behavior changes.

**Comment by magnetophon** [comment]
When I do ``kill $(pidof polybar) `` followed by ``polybar bar1 `` the icons will re-appear after a hide, but only on that one workspace.
If I restart polybar be restarting i3, which calls a slight variant of the standard script, the icons won't reappear after a hide.

**Comment by patrick96** [comment]
This should be fixed by #2609 if you are using the new tray module.
